Mathlib/Tactic/TermCongr.lean

```diff
@@ -0,0 +1,695 @@
+    }and right-hand side {indentD rhs}\n{msg}"
+
+/-- If `lhs` or `rhs` is a congruence hole, then process it.
+Only process ones that are at least as new as `mvarCounterSaved`
+since nothing prevents congruence holes from leaking into the local context. -/
+def mkCongrOfCHole? (mvarCounterSaved : Nat) (lhs rhs : Expr) : MetaM (Option CongrResult) := do
+  match cHole? lhs mvarCounterSaved, cHole? rhs mvarCounterSaved with
+  | some (isLhs1, val1, pf1), some (isLhs2, val2, pf2) =>
+    trace[Elab.congr] "mkCongrOfCHole, both holes"
+    unless isLhs1 == true do
+      throwCongrEx lhs rhs "A RHS congruence hole leaked into the LHS"
+    unless isLhs2 == false do
+      throwCongrEx lhs rhs "A LHS congruence hole leaked into the RHS"
+    -- Defeq checks to unify the lhs and rhs congruence holes.
+    unless ← isDefEq (← inferType pf1) (← inferType pf2) do
+      throwCongrEx lhs rhs "Elaborated types of congruence holes are not defeq."
+    if let some (_, lhsVal, _, rhsVal) ← sides? (← inferType pf1) then
+      unless ← isDefEq val1 lhsVal do
+        throwError "Left-hand side of congruence hole is{indentD lhsVal}\n{""
+          }but is expected to be{indentD val1}"
+      unless ← isDefEq val2 rhsVal do
+        throwError "Right-hand side of congruence hole is{indentD rhsVal}\n{""
+          }but is expected to be{indentD val2}"
+    return some <| CongrResult.mk' val1 val2 pf1
+  | some .., none =>
+    throwCongrEx lhs rhs "Right-hand side lost its congruence hole annotation."
+  | none, some .. =>
+    throwCongrEx lhs rhs "Left-hand side lost its congruence hole annotation."
+  | none, none => return none
+
+/-- Walks along both `lhs` and `rhs` simultaneously to create a congruence lemma between them.
+
+Where they are desyncronized, we fall back to the base case (using `CongrResult.mkDefault'`)
+since it's likely due to unification with the expected type,
+from `_` placeholders or implicit arguments being filled in. -/
+partial def mkCongrOf (depth : Nat) (mvarCounterSaved : Nat) (lhs rhs : Expr) :
+    MetaM CongrResult := do
+  trace[Elab.congr] "mkCongrOf: {depth}, {lhs}, {rhs}, {(← mkFreshExprMVar none).mvarId!}"
+  if depth > 1000 then
+    throwError "congr(...) internal error: out of gas"
+  -- Potentially metavariables get assigned as we process congruence holes,
+  -- so instantiate them to be safe. Placeholders and implicit arguments might
+  -- end up with congruence holes, so they indeed might need a nontrivial congruence.
+  let lhs ← instantiateMVars lhs
+  let rhs ← instantiateMVars rhs
+  if let some res ← mkCongrOfCHole? mvarCounterSaved lhs rhs then
+    trace[Elab.congr] "hole processing succeeded"
+    return res
+  if (hasCHole mvarCounterSaved lhs).isNone && (hasCHole mvarCounterSaved rhs).isNone then
+    -- It's safe to fastforward if the lhs and rhs are defeq and have no congruence holes.
+    -- This is more conservative than necessary since congruence holes might be inside proofs.
```

Comment:
Just making sure I understand here, the "since" means that congruence holes inside proofs wouldn't matter, and not fastforwarding in their presence is the "more conservative"?

Reply:
Correct. I changed it to "since congruence holes might *only* be inside proofs" to try to make the meaning clearer